adapters/huaweiads/huaweiads.go

```diff
@@ -450,72 +458,100 @@ func getHuaweiAdsReqJson(request *huaweiAdsRequest, openRTBRequest *openrtb2.Bid
+// native <-> native
+// video  <->  banner, roll, interstitial, rewarded
+func checkAndExtractOpenrtbFormat(adslot30 *adslot30, adtype int32, yourAdtype string, openRTBImp *openrtb2.Imp) error {
 	if openRTBImp.Banner != nil {
-		if openRTBImp.Banner.W != nil && openRTBImp.Banner.H != nil {
-			adslot30.W = *openRTBImp.Banner.W
-			adslot30.H = *openRTBImp.Banner.H
-		}
-		if len(openRTBImp.Banner.Format) != 0 {
-			var formats = make([]format, 0, len(openRTBImp.Banner.Format))
-			for _, f := range openRTBImp.Banner.Format {
-				if f.H != 0 && f.W != 0 {
-					formats = append(formats, format{f.W, f.H})
-				}
-			}
-			adslot30.Format = formats
+		if adtype != banner && adtype != interstitial {
+			return errors.New("check openrtb format: request has banner, doesn't correspond to huawei adtype " + yourAdtype)
 		}
+		getBannerFormat(adslot30, openRTBImp)
 	} else if openRTBImp.Native != nil {
-		if err := getNativeFormat(&adslot30, openRTBImp); err != nil {
-			return adslot30, err
+		if adtype != native {
+			return errors.New("check openrtb format: request has native, doesn't correspond to huawei adtype " + yourAdtype)
 		}
+		if err := getNativeFormat(adslot30, openRTBImp); err != nil {
+			return err
+		}
+	} else if openRTBImp.Video != nil {
+		if adtype != banner && adtype != interstitial && adtype != rewarded && adtype != roll {
+			return errors.New("check openrtb format: request has video, doesn't correspond to huawei adtype " + yourAdtype)
+		}
+		if err := getVideoFormat(adslot30, adtype, openRTBImp); err != nil {
+			return err
+		}
+	} else if openRTBImp.Audio != nil {
+		return errors.New("check openrtb format: request has audio, not currently supported")
+	} else {
+		return errors.New("check openrtb format: please choose one of our supported type banner, native, or video")
 	}
+	return nil
+}
 
-	// Currently does not support roll type ads, roll ad need TotalDuration
-	if adtypeLower == "roll" {
-		if openRTBImp.Video != nil && openRTBImp.Video.MaxDuration >= 0 {
-			adslot30.TotalDuration = int32(openRTBImp.Video.MaxDuration)
-		} else {
-			return adslot30, errors.New("GetHuaweiAdsReqAdslot30: MaxDuration is empty when adtype is roll.")
+func getBannerFormat(adslot30 *adslot30, openRTBImp *openrtb2.Imp) {
+	if openRTBImp.Banner.W != nil && openRTBImp.Banner.H != nil {
+		adslot30.W = *openRTBImp.Banner.W
+		adslot30.H = *openRTBImp.Banner.H
+	}
+	if len(openRTBImp.Banner.Format) != 0 {
+		var formats = make([]format, 0, len(openRTBImp.Banner.Format))
+		for _, f := range openRTBImp.Banner.Format {
+			if f.H != 0 && f.W != 0 {
+				formats = append(formats, format{f.W, f.H})
+			}
 		}
+		adslot30.Format = formats
 	}
-	return adslot30, nil
 }
 
 func getNativeFormat(adslot30 *adslot30, openRTBImp *openrtb2.Imp) error {
 	if openRTBImp.Native.Request == "" {
-		return errors.New("extractAdmNative: imp.Native.Request is empty")
+		return errors.New("getNativeFormat: imp.Native.Request is empty")
```

Comment:
I'd recommend not using the function name here in the error message.

Reply:
corrected